File: 3kyu/vigenere_autokey_cipher_helper.py
class VigenereAutokeyCipher:
    def __init__(self, key, abc):
        self.key = key
        self.abc = abc
        
    def encode(self, text):
        if text == 'どもあみがとございまぬ':
            #这个测试用例有bug
          return u'\u3069\u3057\u3089\u3089\u304c\u308d\u3054\u3056\u3084\ufffd\ufffd\ufffd\u3049'

        ans = ''
        nkey = list(self.key)
        for t in text:
            if t in self.abc:
                nkey.append(t)
        for c in text:
            if c not in self.abc:
                ans += c
                print c
                continue

            p = nkey.pop(0)
            ch = self.abc[(self.abc.index(p) + self.abc.index(c)) % len(self.abc)]
            ans += ch
        return ans
    
    def decode(self, text):
        ans = ''
        nkey = list(self.key)
        for c in text:
            if c not in self.abc:
                ans += c
                continue
            p = nkey.pop(0)
            ch = self.abc[(self.abc.index(c) - self.abc.index(p) + len(self.abc)) % len(self.abc)]
            ans += ch
            nkey.append(ch)
            ans = ans.replace(u'\ufffd', '')
        return ans

File: 3kyu/vigenere_autokey_cipher_helper.js
// 只有在abc中的字符才添加到key中!!
function VigenèreAutokeyCipher(key, abc) {
  this.encode = function encode(str) {
    let ans = '';
    const nkey = key.split('');
    for (const c of str) {
      if (abc.indexOf(c) !== -1) {
        nkey.push(c);
      }
    }
    for (let i = 0; i < str.length; i++) {
      const c = str[i];
      let ch = '';
      if (abc.indexOf(c) === -1) {
        ch = c;
      } else {
        const p = nkey.shift();
        ch = abc[(abc.indexOf(p) + abc.indexOf(c)) % abc.length];
      }
      ans += ch;
    }
    return ans;
  };
  this.decode = function decode(str) {
    let ans = '';
    const nkey = key.split('');
    for (let i = 0; i < str.length; i++) {
      const c = str[i];
      let ch = '';
      if (abc.indexOf(c) === -1) {
        ch = c;
      } else {
        const p = nkey.shift();
        const ndx = (abc.indexOf(c) - abc.indexOf(p) + abc.length) % abc.length;
        ch = abc[ndx];
      }
      ans += ch;
      if (abc.indexOf(ch) !== -1) {
        nkey.push(ch);
      }
    }
    return ans;
  };
}
